src/pcap/api.go

```diff
@@ -0,0 +1,650 @@
+		for n := range c {
+			out <- n
+		}
+		wg.Done()
+	}
+	wg.Add(len(cs))
+	for _, c := range cs {
+		go output(c)
+	}
+
+	// Start a goroutine to close out once all the output goroutines are done.
+	go func() {
+		wg.Wait()
+		close(out)
+	}()
+	return out
+}
+
+type streamPrep struct {
+}
+
+// prepareStreamToTarget creates a client connection to the given target, contacts the client API for the Agent service
+// to start the Capture.
+func (p *streamPrep) prepareStreamToTarget(ctx context.Context, req *CaptureOptions, target AgentEndpoint, creds credentials.TransportCredentials, log *zap.Logger) (captureReceiver, error) {
+	cc, err := grpc.Dial(target.String(), grpc.WithTransportCredentials(creds))
+	if err != nil {
+		err = fmt.Errorf("start capture from '%s': %w", target, err)
+		return nil, err
+	}
+
+	agent := NewAgentClient(cc)
+
+	statusRes, err := agent.Status(ctx, &StatusRequest{})
+	err = checkAgentStatus(statusRes, err, target)
+	if err != nil {
+		return nil, err
+	}
+
+	agentContext := context.Background()
+	vcapID, err := vcapIDFromOutgoingCtx(ctx)
+	if err != nil {
+		// TODO impelement error handling
+	}
+	agentContext, log = setVcapID(agentContext, log, vcapID)
+	captureStream, err := agent.Capture(agentContext)
+
+	if err != nil {
+		convertStatusCodeToMsg(err, target)
+		return nil, err
+	}
+
+	patchedFilter, err := patchFilter(req.Filter)
+	if err != nil {
+		err = pcapError{
+			status: status.New(codes.FailedPrecondition, "Expanding the pcap filter to exclude traffic to pcap-api failed"),
+			inner:  err,
+		}
+		convertStatusCodeToMsg(err, target)
+		return nil, err
+	}
+	req.Filter = patchedFilter
+
+	err = captureStream.Send(&AgentRequest{
+		Payload: &AgentRequest_Start{
+			Start: &StartAgentCapture{
+				Capture: req,
+			},
+		},
+	})
+	if err != nil {
+		// out <- convertStatusCodeToMsg(err, target)
+		return nil, err
+	}
+	return captureStream, nil
+}
+
+type captureReceiver interface {
+	Recv() (*CaptureResponse, error)
+	Send(*AgentRequest) error
+	CloseSend() error
+	Context() context.Context
+}
+
+// readMsgFromStream reads Capture messages from stream and outputs them to the out channel.If the given context errors
+// an AgentRequest_Stop is sent and the messages continue to be read.if context will be cancelled from other routine
+// (mostly  because client requests to stop capture), the stop request will be forwarded to agent. The data from the agent will be read till stream ends with EOF.
+func readMsgFromStream(ctx context.Context, captureStream captureReceiver, target AgentEndpoint, bufSize int) <-chan *CaptureResponse {
+	out := make(chan *CaptureResponse, bufSize)
+	stopped := false
+	go func() {
+		defer close(out)
+		// defer wg.Done()
+		defer func(captureStream captureReceiver) {
+			closeSendErr := captureStream.CloseSend()
+			if closeSendErr != nil {
+				out <- convertStatusCodeToMsg(closeSendErr, target)
+				return
+			}
+		}(captureStream)
+		for {
+			if ctx.Err() != nil && !stopped {
```

Comment:
Now I'm completely lost. From what I can see there is control via context cancellation. Why do we then have the draining logic with all the locks etc.?

Reply:
The drain logic has beed enhanced.